### ISSUE TITLE
Vim9: not easy to find where whitespace is disallowed

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#7791]
**Describe the bug**

In Vim9 script, it is not always easy to find where a whitespace is disallowed.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9
        def Func(l: any)
        enddef
        Func(['a', 'd', 'gj', 'gqq' , 'i', 'o', 'r', 'u', 'x'])
    EOF
    )

`E1068` is raised:

    E1068: No white space allowed before ','

This is expected; but no context is given about where this superfluous whitespace is written.

**Expected behavior**

The error message gives us more context about where the superfluous whitespace is written.

**Environment**

 - Vim version: 8.2 Included patches: 1-2475
 - OS: Ubuntu 16.04.7 LTS
 - Terminal: xterm(365)

**Additional context**

The superfluous whitespace is here:

    Func(['a', 'd', 'gj', 'gqq' , 'i', 'o', 'r', 'u', 'x'])
                               ^
                               ✘

Therefore, this error message would be more helpful:

    E1068: No white space allowed before ',' at ", 'i', 'o', 'r', 'u', 'x'"
                                             ^----------------------------^

A similar issue was fixed in [8.2.2266](https://github.com/vim/vim/commit/e7a73e07625b64a40671a0007ad38a34cbe9d1ee).  I tried to copy the relevant code, and came up with this patch:
```diff
diff --git a/src/dict.c b/src/dict.c
index 260f22985..17958189a 100644
--- a/src/dict.c
+++ b/src/dict.c
@@ -945,7 +945,7 @@ eval_dict(char_u **arg, typval_T *rettv, evalarg_T *evalarg, int literal)
 	if (**arg != ':')
 	{
 	    if (*skipwhite(*arg) == ':')
-		semsg(_(e_no_white_space_allowed_before_str), ":");
+		semsg(_(e_no_white_space_allowed_before_str_at_str), ":", *arg);
 	    else
 		semsg(_(e_missing_dict_colon), *arg);
 	    clear_tv(&tvkey);
@@ -1025,7 +1025,7 @@ eval_dict(char_u **arg, typval_T *rettv, evalarg_T *evalarg, int literal)
 	if (!had_comma)
 	{
 	    if (**arg == ',')
-		semsg(_(e_no_white_space_allowed_before_str), ",");
+		semsg(_(e_no_white_space_allowed_before_str_at_str), ",", *arg);
 	    else
 		semsg(_(e_missing_dict_comma), *arg);
 	    goto failret;
diff --git a/src/errors.h b/src/errors.h
index e6d7ab26c..d3642766b 100644
--- a/src/errors.h
+++ b/src/errors.h
@@ -173,8 +173,8 @@ EXTERN char e_cannot_declare_a_register_str[]
 	INIT(= N_("E1066: Cannot declare a register: %s"));
 EXTERN char e_separator_mismatch_str[]
 	INIT(= N_("E1067: Separator mismatch: %s"));
-EXTERN char e_no_white_space_allowed_before_str[]
-	INIT(= N_("E1068: No white space allowed before '%s'"));
+EXTERN char e_no_white_space_allowed_before_str_at_str[]
+	INIT(= N_("E1068: No white space allowed before '%s' at \"%s\""));
 EXTERN char e_white_space_required_after_str[]
 	INIT(= N_("E1069: White space required after '%s'"));
 EXTERN char e_missing_from[]
diff --git a/src/list.c b/src/list.c
index c51c99bc9..449533e0b 100644
--- a/src/list.c
+++ b/src/list.c
@@ -1328,7 +1328,7 @@ eval_list(char_u **arg, typval_T *rettv, evalarg_T *evalarg, int do_error)
 	    if (do_error)
 	    {
 		if (**arg == ',')
-		    semsg(_(e_no_white_space_allowed_before_str), ",");
+		    semsg(_(e_no_white_space_allowed_before_str_at_str), ",", *arg);
 		else
 		    semsg(_("E696: Missing comma in List: %s"), *arg);
 	    }
diff --git a/src/userfunc.c b/src/userfunc.c
index bdf406426..bab213a6f 100644
--- a/src/userfunc.c
+++ b/src/userfunc.c
@@ -873,7 +873,7 @@ get_func_tv(
 	{
 	    if (*argp != ',' && *skipwhite(argp) == ',')
 	    {
-		semsg(_(e_no_white_space_allowed_before_str), ",");
+		semsg(_(e_no_white_space_allowed_before_str_at_str), ",", argp);
 		ret = FAIL;
 		break;
 	    }
@@ -3209,7 +3209,7 @@ define_function(exarg_T *eap, char_u *name_arg)
 
     if ((vim9script || eap->cmdidx == CMD_def) && VIM_ISWHITE(p[-1]))
     {
-	semsg(_(e_no_white_space_allowed_before_str), "(");
+	semsg(_(e_no_white_space_allowed_before_str_at_str), "(", p);
 	goto ret_free;
     }
 
diff --git a/src/vim9compile.c b/src/vim9compile.c
index 6a1d8d9e7..77f0fb6e1 100644
--- a/src/vim9compile.c
+++ b/src/vim9compile.c
@@ -2800,7 +2800,7 @@ compile_arguments(char_u **arg, cctx_T *cctx, int *argcount)
 
 	if (*p != ',' && *skipwhite(p) == ',')
 	{
-	    semsg(_(e_no_white_space_allowed_before_str), ",");
+	    semsg(_(e_no_white_space_allowed_before_str_at_str), ",", p);
 	    p = skipwhite(p);
 	}
 	if (*p == ',')
@@ -3055,7 +3055,7 @@ compile_list(char_u **arg, cctx_T *cctx, ppconst_T *ppconst)
 	}
 	if (*p == ',')
 	{
-	    semsg(_(e_no_white_space_allowed_before_str), ",");
+	    semsg(_(e_no_white_space_allowed_before_str_at_str), ",", p);
 	    return FAIL;
 	}
 	if (*p == ']')
@@ -3234,7 +3234,7 @@ compile_dict(char_u **arg, cctx_T *cctx, ppconst_T *ppconst)
 	if (**arg != ':')
 	{
 	    if (*skipwhite(*arg) == ':')
-		semsg(_(e_no_white_space_allowed_before_str), ":");
+		semsg(_(e_no_white_space_allowed_before_str_at_str), ":", *arg);
 	    else
 		semsg(_(e_missing_dict_colon), *arg);
 	    return FAIL;
@@ -3273,7 +3273,7 @@ compile_dict(char_u **arg, cctx_T *cctx, ppconst_T *ppconst)
 	}
 	if (IS_WHITE_OR_NUL(*whitep))
 	{
-	    semsg(_(e_no_white_space_allowed_before_str), ",");
+	    semsg(_(e_no_white_space_allowed_before_str_at_str), ",", *arg);
 	    return FAIL;
 	}
 	whitep = *arg + 1;
@@ -4270,7 +4270,7 @@ compile_expr7t(char_u **arg, cctx_T *cctx, ppconst_T *ppconst)
 	if (**arg != '>')
 	{
 	    if (*skipwhite(*arg) == '>')
-		semsg(_(e_no_white_space_allowed_before_str), ">");
+		semsg(_(e_no_white_space_allowed_before_str_at_str), ">", *arg);
 	    else
 		emsg(_(e_missing_gt));
 	    return FAIL;
diff --git a/src/vim9type.c b/src/vim9type.c
index 1374cee1b..814245f18 100644
--- a/src/vim9type.c
+++ b/src/vim9type.c
@@ -638,7 +638,7 @@ parse_type_member(
 	if (give_error)
 	{
 	    if (*skipwhite(*arg) == '<')
-		semsg(_(e_no_white_space_allowed_before_str), "<");
+		semsg(_(e_no_white_space_allowed_before_str_at_str), "<", *arg);
 	    else
 		emsg(_(e_missing_type));
 	}
@@ -779,7 +779,7 @@ parse_type(char_u **arg, garray_T *type_gap, int give_error)
 			if (*p != ',' && *skipwhite(p) == ',')
 			{
 			    if (give_error)
-				semsg(_(e_no_white_space_allowed_before_str), ",");
+				semsg(_(e_no_white_space_allowed_before_str_at_str), ",", p);
 			    return NULL;
 			}
 			if (*p == ',')
```
It seems to fix the issue.  I ran the tests and they all passed (a few were skipped); which is weird, I would have expected a few tests to fail considering that the error messages are now sometimes different...

**Comment by lacygoill** [comment]
Probably fixed by [8.2.2480](https://github.com/vim/vim/releases/tag/v8.2.2480).

**Comment by lacygoill** [comment]
Actually no, it was fixed by [8.2.2486](https://github.com/vim/vim/releases/tag/v8.2.2486).
